### PR TITLE
Add torch.cuda rng state to seed save/load

### PR DESCRIPTION
**Pull request by Anner-deJong**
## What does this PR do?

Add torch.cuda rng state to the default seed save/load functionality:
- [Confirmed through testing on dev GPU server] `torch.get/set_rng_state()` does not include `torch.cuda` * and `torch.cuda` rng is not reset
- Confirmed manually on a dev GPU server the test should do what it is supposed, actual branch + dev is on a macbook without gpu, I presume the github merge tests will cover

\* Presumably cause `torch.cuda` is not included in the torch `default_generator` `.set_state()` and `.get_state()` (C implementation [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/Generator.cpp))

Fixes #UNKNOWN

### Does your PR introduce any breaking changes? If yes, please list them.

Not aware

## Before submitting

- [No] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [Yes] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [Yes] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [Yes] Did you make sure to **update the documentation** with your changes? (if necessary)
- [Yes] Did you write any **new necessary tests**? (not for typos and docs)
- [Yes] Did you verify new and **existing tests pass** locally with your changes?
- [UNAWARE OF ANY] Did you list all the **breaking changes** introduced by this pull request?
- [Yes] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [Yes] Is this pull request ready for review? (if not, please submit in draft mode)
- [Mostly] Check that all items from **Before submitting** are resolved
- [Yes] Make sure the title is self-explanatory and the description concisely explains the PR
- [Added by @awaelchli, thanks!] Add labels and milestones (and optionally projects) to the PR so it can be classified

#### YES LIGHTNING IS AWESOME ⚡ 

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Anner-deJong** [comment]
Only issue I can come up with if cuda is not available, torch.cuda returns None?
Found the tests (I forked too small part of the repo, will check + maybe update)
Putting PR in draft mode for the time being

**Comment by Anner-deJong** [comment]
> Only issue I can come up with if cuda is not available, torch.cuda returns None? Found the tests (I forked too small part of the repo, will check + maybe update) Putting PR in draft mode for the time being

Updated the test and did some interactive testing, ready for review. If torch is not compiled with cuda, torch.cuda.get/set.. doesnt run into any error. On the test side, `torch.cuda.FloatTensor(2).normal_()`, which is not guarded by an `torch.cuda.is_available()`

File: src/pytorch_lightning/utilities/seed.py [hunk 1]
```diff
@@ -121,13 +121,19 @@ def pl_worker_init_function(worker_id: int, rank: Optional[int] = None) -> None:
...
+        "torch.cuda": torch.cuda.get_rng_state_all(),
+        "numpy": np.random.get_state(),
+        "python": python_get_rng_state(),
+    }
 
 
 def _set_rng_states(rng_state_dict: Dict[str, Any]) -> None:
     """Set the global random state of :mod:`torch`, :mod:`numpy` and Python in the current process."""
     torch.set_rng_state(rng_state_dict["torch"])
+    torch.cuda.set_rng_state_all(rng_state_dict["torch.cuda"])
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/seed.py:136]: I'm wondering, how big is this state (e.g. in bytes)?

File: src/pytorch_lightning/utilities/seed.py [hunk 2]
```diff
@@ -121,13 +121,19 @@ def pl_worker_init_function(worker_id: int, rank: Optional[int] = None) -> None:
 
 
 def _collect_rng_states() -> Dict[str, Any]:
-    """Collect the global random state of :mod:`torch`, :mod:`numpy` and Python."""
-    return {"torch": torch.get_rng_state(), "numpy": np.random.get_state(), "python": python_get_rng_state()}
+    """Collect the global random state of :mod:`torch`, :mod:`torch.cuda`,:mod:`numpy` and Python."""
+    return {
+        "torch": torch.get_rng_state(),
+        "torch.cuda": torch.cuda.get_rng_state_all(),
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/seed.py:127]: Returns empty list on non-cuda enabled device. LGTM!

**Review by awaelchli — APPROVED** [review]
Great!

**Comment by awaelchli** [comment]
Could you also add a changelog entry? <3

File: src/pytorch_lightning/utilities/seed.py (hunk 1 above)
Comment (Anner-deJong) [reply src/pytorch_lightning/utilities/seed.py:136]: `sys.getsizeof` reports me 88 bytes

(torch.get_rng_state() -> 72 bytes, and both sizes remain the same after
```
for _ in range(10000000):
        torch.rand(3, 2)
        torch.cuda.FloatTensor(3, 2).normal_()
```

**Comment by Anner-deJong** [comment]
> Could you also add a changelog entry? <3

Done!

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/utilities/seed.py (hunk 1 above)
Comment (carmocca) [inline src/pytorch_lightning/utilities/seed.py:136]: This should avoid KeyErrors for support with old `rng_state_dict`s

Comment (awaelchli) [reply src/pytorch_lightning/utilities/seed.py:136]: Thanks! Looks reasonable, so no problem there :)

Comment (Anner-deJong) [reply src/pytorch_lightning/utilities/seed.py:136]: Good catch, updating now

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md [hunk 3]
```diff
@@ -29,6 +29,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Enabled `on_before_batch_transfer` for `DPStrategy` and `IPUAccelerator` ([14023](https://github.com/Lightning-AI/lightning/pull/14023))
 
+- Included `torch.cuda` rng state to the aggregate `_collect_rng_states()` and `_set_rng_states()` ([14384](https://github.com/Lightning-AI/lightning/pull/14384))
```
Comment (rohitgr7) [inline src/pytorch_lightning/CHANGELOG.md:32]:
```suggestion
- Included `torch.cuda` rng state to the aggregate `_collect_rng_states()` and `_set_rng_states()` ([#14384](https://github.com/Lightning-AI/lightning/pull/14384))
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 4]
```diff
@@ -29,6 +29,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Enabled `on_before_batch_transfer` for `DPStrategy` and `IPUAccelerator` ([14023](https://github.com/Lightning-AI/lightning/pull/14023))
```
Comment (Anner-deJong) [inline src/pytorch_lightning/CHANGELOG.md:30]:
```suggestion
- Enabled `on_before_batch_transfer` for `DPStrategy` and `IPUAccelerator` ([#14023](https://github.com/Lightning-AI/lightning/pull/14023))
```
